docs/proposals/20221611-workload-identity-integration.md

```diff
@@ -0,0 +1,351 @@
+
+```yaml
+kind: Cluster
+apiVersion: kind.x-k8s.io/v1alpha4
+nodes:
+- role: control-plane
+  extraMounts:
+      # path on node where the public key exists
+    - hostPath: ${SERVICE_ACCOUNT_KEY_FILE}
+      containerPath: /etc/kubernetes/pki/sa.pub
+	  # path on node where the private key exists
+    - hostPath: ${SERVICE_ACCOUNT_SIGNING_KEY_FILE}
+      containerPath: /etc/kubernetes/pki/sa.key
+  kubeadmConfigPatches:
+  - |
+    kind: ClusterConfiguration
+    apiServer:
+      extraArgs:
+	    # the oidc url after it has been set up
+        service-account-issuer: ${SERVICE_ACCOUNT_ISSUER}
+        service-account-key-file: /etc/kubernetes/pki/sa.pub
+        service-account-signing-key-file: /etc/kubernetes/pki/sa.key
+    controllerManager:
+      extraArgs:
+        service-account-private-key-file: /etc/kubernetes/pki/sa.key
+```
+
+#### <a name='FederatedCredential'></a>Federated Credential
+
+The service account that the capz pod will use should be annotated with the client ID of the AAD application or User Assigned Identity and a fedrated identity should be created by associating the ServiceAccont as a subject.
+
+```yaml
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  annotations:
+    azure.workload.identity/client-id: ${APPLICATION_CLIENT_ID:-$USER_ASSIGNED_IDENTITY_CLIENT_ID}
+  labels:
+    azure.workload.identity/use: "true"
+  name: ${SERVICE_ACCOUNT_NAME}
+  namespace: ${SERVICE_ACCOUNT_NAMESPACE}
+```
+
+Also, a federated identity should be created using the azure cli ( or via the azure portal).
+
+```bash
+az identity federated-credential create \
+  --name "kubernetes-federated-credential" \
+  --identity-name "${USER_ASSIGNED_IDENTITY_NAME}" \
+  --resource-group "${RESOURCE_GROUP}" \
+  --issuer "${SERVICE_ACCOUNT_ISSUER}" \
+  --subject "system:serviceaccount:${SERVICE_ACCOUNT_NAMESPACE}:${SERVICE_ACCOUNT_NAME}"
+```
+
+This means there are two approaches to associate the correct service account to the capz deployment.
+
+- Deploy the capz with the usual configuration YAML and then patch the service account with the client ID.
+
+- Or create the service account prior to capz deployment which uses the same service account.
+
+
+#### <a name='DistributeKeys'></a>Distribute Keys To Management Cluster
+
+Setting up credentials for management cluster which is created from the bootstrap cluster is key activity. This requires storing the keys on control plane node. One approach to do so can be the following:
+
+- A reular `clusterctl init` is done.
+- Once the cluster is up and running, the keys can be distributed on the control plane node via scp.
+- After distributing the key, the appropriate flags as it was set up for the bootstrap cluster can be setup.
+- The service account should be annotated with the client ID.
+- At this stage, `clusterctl move` can be executed.
+
+Above activities can be automated via clusterctl subcommands that is specific to azure infrastructre.
```

Comment:
not sure I understand this part

Reply:
One of the usual workflows that I have seen from users is to use Bootstrap & Pivot : 
- Creating a kind cluster and using it as a bootstrap cluster to deploy a workload cluster on Azure. 
- Now that workload cluster is converted to a management cluster using the pivot process. 

More on this link
https://cluster-api.sigs.k8s.io/clusterctl/commands/move.html?highlight=pivot#bootstrap--pivot

Now with azwi, this means making the key pairs available non the control plane nodes and doing an extraMount as we are doing it for the bootstrap cluster.  

With that I was thinking on the lines for making a subcommand that can deal with such thing e.g. transfer the keys to all the control plane nodes via scp( as that looks a manual process that is now introduces to be able to use azwi.)